src/manifolds/Symplectic.jl

```diff
@@ -0,0 +1,889 @@
+    U_inv = lu(add_scaled_I!(symplectic_inverse_times(M, p, q), 1))
+    V_inv = lu(add_scaled_I!(symplectic_inverse_times(M, q, p), 1))
+
+    X .= 2 .* ((p / V_inv .- p / U_inv) .+ ((p + q) / U_inv) .- p)
+    return X
+end
+
+@doc raw"""
+    manifold_dimension(::Symplectic{n})
+
+Returns the dimension of the symplectic manifold
+embedded in ``ℝ^{2n \times 2n}``, i.e.
+````math
+    \operatorname{dim}(\operatorname{Sp}(2n)) = (2n + 1)n.
+````
+"""
+manifold_dimension(::Symplectic{n}) where {n} = (2n + 1) * n
+
+@doc raw"""
+    project_riemannian!(M::Symplectic{n, ℝ}, Y, p, X) where {n}
+
+Compute the projection of ``X ∈ R^{2n × 2n}`` onto ``T_p\operatorname{Sp}(2n, ℝ)`` w.r.t.
+the Riemannian metric ``g`` [`RealSymplecticMetric`](@ref).
+
+The closed form projection mapping is given by [^Gao2021riemannian]
+````math
+    \operatorname{P}^{T_p\operatorname{Sp}(2n)}_{g_p}(X) = pQ\operatorname{sym}(p^TQ^TX),
+````
+where ``\operatorname{sym}(A) = \frac{1}{2}(A + A^T)``.
+"""
+function project_riemannian!(::Symplectic, Y, p, X)
+    Q = SymplecticMatrix(p, X)
+
+    pT_QT_X = p' * Q' * X
+    symmetrized_pT_QT_X = (1 / 2) .* (pT_QT_X + pT_QT_X')
+
+    Y[:, :] = p * Q * (symmetrized_pT_QT_X)
+    return Y
+end
+
+@doc raw"""
+    project_riemannian_normal!(M::Symplectic{n, ℝ}, Y, p, X)
+
+Project onto the normal of the tangent space ``(T_p\operatorname{Sp}(2n))^{\perp_g}`` at
+a point ``p ∈ \operatorname{Sp}(2n)``,
+relative to the riemannian metric ``g`` [`RealSymplecticMetric`](@ref).
+That is,
+````math
+(T_p\operatorname{Sp}(2n))^{\perp_g} = \{Y \in \mathbb{R}^{2n \times 2n} :
+                        g_p(Y, X) = 0 \;\forall\; X \in T_p\operatorname{Sp}(2n)\}.
+````
+The closed form projection operator onto the normal space is given by [^Gao2021riemannian]
+````math
+\operatorname{P}^{(T_p\operatorname{Sp}(2n))\perp}_{g_p}(X) = pQ\operatorname{skew}(p^TQ^TX),
+````
+where ``\operatorname{skew}(A) = \frac{1}{2}(A - A^T)``.
+
+[^Gao2021riemannian]:
+    > Gao, Bin and Son, Nguyen Thanh and Absil, P-A and Stykel, Tatjana:
+    > Riemannian optimization on the symplectic Stiefel manifold,
+    > SIAM Journal on Optimization 31(2), pp. 1546-1575, 2021.
+    > doi [10.1137/20M1348522](https://doi.org/10.1137/20M1348522)
+"""
+function project_riemannian_normal!(::Symplectic, Y, p, X)
+    Q = SymplecticMatrix(p, X)
+
+    pT_QT_X = p' * Q' * X
+    skew_pT_QT_X = (1 / 2) .* (pT_QT_X .- pT_QT_X')
+
+    Y[:, :] = p * Q * skew_pT_QT_X
+    return Y
+end
+
+function Base.rand(M::Symplectic{n}, hamiltonian_norm::Number=1.0) where {n}
+    # Generate random matrices to construct a Hamiltonian matrix:
+    Ω = rand_hamiltonian(M; final_norm=hamiltonian_norm)
+    # Return 'cay(Ω)':
+    return (I - Ω) \ (I + Ω)
+end
+
+function Base.rand(::Symplectic{n}, p::AbstractMatrix) where {n}
+    # Generate random symmetric matrix:
+    S = rand(2n, 2n) .- 1 / 2
+    S .= (1 / 2) .* (S + S')
+    Q = SymplecticMatrix(p)
+    lmul!(Q, S)
+    return p * S
+end
```

Comment:
This API is confusing, maybe let's make this a separate function instead of adding a method to `Base.rand`? For example `random_symmetric_matrix`.

Reply:
`random_vector(::Symplectic, p)` should be fine I think, it follows our naming for other functions that operate on tangent vectors.